flake.nix

```
@@ -54,7 +54,7 @@
           - `system`: Legacy alias for `nixpkgs.hostPlatform`, but this is already set in the generated `hardware-configuration.nix`, included by `configuration.nix`.
           - `pkgs`: Legacy alias for `nixpkgs.pkgs`; use `nixpkgs.pkgs` and `nixosModules.readOnlyPkgs` instead.
         */
-        nixosSystem = args:
+        nixosSystem = args: final.warnIf (args?specialArgs && args.specialArgs?pkgs && !(args?allowPkgMasking && args.allowPkgMasking)) "nixosSystem: Masking pkgs in specialArgs is not recommended, as it can lead to unexpected behavior. Use overlays instead. If this is really the behavior you want, set allowPkgMasking = true in nixosSystem."
```

This applies to all entrypoints, so it'd be better to do this in `misc/nixpkgs.nix`.

---

```suggestion
        nixosSystem = args: final.warnIf (args?specialArgs && args.specialArgs?pkgs && !(args?allowPkgMasking && args.allowPkgMasking)) "You have set specialArgs.pkgs, which means that options like nixpkgs.config and nixpkgs.overlays will be ignored. If you wish to reuse an already created pkgs, which you know is configured correctly for this NixOS configuration, please import the `nixosModules.pkgsReadOnly` module from the nixpkgs flake or `(modulesPath + "/misc/nixpkgs/read-only.nix"), and set `{ nixpkgs.pkgs = <your pkgs>; }`. This properly disables the ignored options to prevent future surprises."
```

Reusing a `pkgs` across configurations is a valid strategy for making sure the same package set is used between multiple configuration and/or other things in a project. It can also improve evaluation performance when those things are evaluated in one go.